### PR TITLE
feat: ERC20 token bridging

### PR DESCRIPTION
**Pull request by kupermind**
- ERC20 token bridging from Polygon to ETH.

### CODE REVIEW

File: test/FxERC20.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+            await expect(
+                FxERC20ChildTunnel.deploy(signers[1].address, signers[1].address, AddressZero)
+            ).to.be.revertedWithCustomError(fxERC20ChildTunnel, "ZeroAddress");
+        });
+    });
+
+    context("Deposit and withdraw ERC20 tokens", async function () {
+        it("Should fail when trying to call from incorrect contract addresses", async function () {
+            // FxChild as a sender is incorrect
```
Comment (mariapiamo) [inline test/FxERC20.js:86]:
```suggestion
            // Deployer as a sender is incorrect
```

File: test/FxERC20.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            await expect(
+                fxERC20ChildTunnel.connect(signers[1]).processMessageFromRoot(stateId, deployer.address, "0x")
+            ).to.be.revertedWith("FxBaseChildTunnel: INVALID_SENDER");
+
+            // FxRoot as a sender is incorrect
```
Comment (mariapiamo) [inline test/FxERC20.js:91]:
```suggestion
            // signers[1].address as a sender is incorrect
           // FxChild as a sender is correct
```

File: test/FxERC20.js (hunk 1 above)
Comment (mariapiamo) [reply test/FxERC20.js:86]: fxERC20RootTunnel is the correct one

File: test/FxERC20.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            // Withdraw tokens
+            await rootToken.connect(deployer).approve(fxERC20RootTunnel.address, amount);
+
+            // Root token must be owned by the FxERC20RootTunnel contract
+            await rootToken.changeOwner(fxERC20RootTunnel.address);
+
+            // Burn tokens on L1 and send message to L2 to retrieve them there
+            await expect(
+                fxERC20RootTunnel.withdraw(amount)
+            ).to.be.reverted;
```
Comment (mariapiamo) [inline test/FxERC20.js:152]: why?

File: test/FxERC20.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                ["address", "address", "uint256"],
+                [deployer.address, deployer.address, amount]
+            );
+
+            // Upon message receive, tokens on L2 are transferred to the destination account (deployer)
+            await fxERC20ChildTunnel.connect(deployer).processMessageFromRoot(stateId, fxERC20RootTunnel.address, data);
+
+            // There must be no balance left locked on the FxERC20ChildTunnel contract
+            balance = await childToken.balanceOf(fxERC20ChildTunnel.address);
+            expect(balance).to.equal(0);
```
Comment (mariapiamo) [inline test/FxERC20.js:165]: where we check that the receiver has non-zero balance?

File: test/FxERC20.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
```
Comment (mariapiamo) [inline test/FxERC20.js:206]: same comment above

File: contracts/bridges/FxERC20ChildTunnel.sol [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+/// @dev Zero value when it has to be different from zero.
+error ZeroValue();
+
+/// @title FxERC20ChildTunnel - Smart contract for the L2 token management part
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract FxERC20ChildTunnel is FxBaseChildTunnel {
+    event FxDepositERC20(address indexed childToken, address indexed rootToken, address from, address indexed to, uint256 amount);
```
[the remark this answers is not in the document]
Comment (kupermind) [reply contracts/bridges/FxERC20ChildTunnel.sol:18]: We can only index 3 variables in a single event. `from` is less important than `to`

File: contracts/bridges/FxERC20RootTunnel.sol [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /// @dev Withdraws bridged tokens on L1 in order to obtain their original version on L2.
+    /// @notice Destination address is the same as the sender address.
+    /// @param amount Token amount to be withdrawn.
+    function withdraw(uint256 amount) external {
+        _withdraw(msg.sender, amount);
+    }
+
+    /// @dev Withdraws bridged tokens on L1 in order to obtain their original version on L2 by a specified address.
```
[the remark this answers is not in the document]
Comment (kupermind) [reply contracts/bridges/FxERC20RootTunnel.sol:51]: obtain on L2 by the address, or obtain on L2 with the address, or we choose a different wording :)

File: scripts/deployment/bridges/polygon/deploy_03_erc20_child_tunnel.js
[the remark this answers is not in the document]
Comment (kupermind) [reply scripts/deployment/bridges/polygon/deploy_03_erc20_child_tunnel.js:1]: They go in different folders as they are on different chains, I am not sure what would be the best approach here.

File: test/FxERC20.js (hunk 3 above)
Comment (kupermind) [reply test/FxERC20.js:152]: The `FxERC20RootTunnel` ultimately leads to the `FxRoot` contract calling the cross-chain `StatSync`, which is impossible to recreate in a testing environment. I can try to write an `FxRootMock` contract for that purpose.

File: test/FxERC20.js (hunk 4 above)
Comment (kupermind) [reply test/FxERC20.js:165]: was just lazy to check in one place :P

File: contracts/bridges/FxERC20RootTunnel.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (mariapiamo) [reply contracts/bridges/FxERC20RootTunnel.sol:51]: ah now I see, I had in mind you wanted to specify that here it is withdrawn to a specific address.

File: test/FxERC20.js (hunk 1 above)
Comment (kupermind) [reply test/FxERC20.js:86]: Here it's the `FxChild`, which is `deployer.address` in this test setup, and not the `signers[1].address.` Corrected.

File: test/FxERC20.js (hunk 3 above)
Comment (kupermind) [reply test/FxERC20.js]: FxRootMock is in place

**Review by mariapiamo — APPROVED** [review]

**Pull request merged**